Dualstack downward api

aramase opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test

/kind feature
> /kind flake

**What this PR does / why we need it**:

https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20180612-ipv4-ipv6-dual-stack.md#container-environment-variables

- adds downward api support for `podIPs`
- updates the host file with v4 and v6 pod IPs for dualstack

```    
        env:
        - name: MY_POD_IPS
          valueFrom:
            fieldRef:
              fieldPath: status.podIPs
```
```
$ kubectl exec -ti nginx-8bff7fdb7-xdzj6 -- env | grep POD_IPS
MY_POD_IPS=10.244.0.6,fd00::6
```
```
$ kubectl exec -ti nginx-8bff7fdb7-xdzj6 -- cat /etc/hosts
# Kubernetes-managed hosts file.
127.0.0.1    localhost
::1    localhost ip6-localhost ip6-loopback
fe00::0    ip6-localnet
fe00::0    ip6-mcastprefix
fe00::1    ip6-allnodes
fe00::2    ip6-allrouters
10.244.0.6    nginx-8bff7fdb7-xdzj6
fd00::6    nginx-8bff7fdb7-xdzj6
```

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
The following information is available through environment variables:

status.podIPs - the pod's IP addresses
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
KEP - https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20180612-ipv4-ipv6-dual-stack.md
```

aramase committed:
add status.podIPs in downward api

add host file write for podIPs

update tests

remove import alias

update type check

update type check

remove import alias

update open api spec

add tests

update test

add tests

address review comments

update imports

remove todo and import alias

aramase commented:
/test pull-kubernetes-bazel-build

aojea commented on pkg/kubelet/kubelet_pods_test.go:
```diff
@@ -319,12 +319,28 @@ fe00::2	ip6-allrouters
+`,
+		},
+		{
+			[]string{"203.0.113.1", "fd00::/6"},
```
shouldn't this be IP addresses only, instead of subnets? 
`fd00::6/` instead of `fd00::/6`

aojea commented on pkg/kubelet/kubelet_pods_test.go:
```diff
@@ -319,12 +319,28 @@ fe00::2	ip6-allrouters
+fe00::1	ip6-allnodes
+fe00::2	ip6-allrouters
+203.0.113.1	podFoo.domainFoo	podFoo
+fd00::/6	podFoo.domainFoo	podFoo
```
ditto `fd00::6` instead of `fd00::/6`

aramase replied:
Thanks @aojea, Updated!

aojea commented on pkg/kubelet/kubelet_pods_test.go:
```diff
@@ -722,6 +747,7 @@ func TestMakeEnvironmentVariables(t *testing.T) {
 				{Name: "POD_NODE_NAME", Value: "node-name"},
 				{Name: "POD_SERVICE_ACCOUNT_NAME", Value: "special"},
 				{Name: "POD_IP", Value: "1.2.3.4"},
+				{Name: "POD_IPS", Value: "1.2.3.4,fd00::/6"},
```
`fd00::6` instead of `fd00::/6`

aojea commented on pkg/kubelet/kubelet_pods_test.go:
```diff
@@ -1627,8 +1653,9 @@ func TestMakeEnvironmentVariables(t *testing.T) {
 			},
 		}
 		podIP := "1.2.3.4"
+		podIPs := []string{"1.2.3.4,fd00::/6"}
```
`fd00::6` instead of `fd00::/6`

aojea commented:
I think this is introducing a user-facing change that has to be explained in the section `Does this PR introduce a user-facing change?:`

aramase commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

khenidak commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -772,7 +772,7 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, podStatus *kubecontaine
 
 		klog.V(4).Infof("Creating %v %+v in pod %v", typeName, container, format.Pod(pod))
-		if msg, err := m.startContainer(podSandboxID, podSandboxConfig, container, pod, podStatus, pullSecrets, podIP); err != nil {
+		if msg, err := m.startContainer(podSandboxID, podSandboxConfig, container, pod, podStatus, pullSecrets, podIP, podIPs); err != nil {
```
Please update the comment https://github.com/kubernetes/kubernetes/pull/83123/files#diff-8da51ab012ecbdf2b013ecbcf1c37033R684

aramase replied:
done!

khenidak commented on pkg/kubelet/kuberuntime/kuberuntime_manager_test.go:
```diff
@@ -156,7 +156,7 @@ func makeFakeContainer(t *testing.T, m *kubeGenericRuntimeManager, template cont
 	assert.NoError(t, err, "generatePodSandboxConfig for container template %+v", template)
 
-	containerConfig, _, err := m.generateContainerConfig(template.container, template.pod, template.attempt, "", template.container.Image)
+	containerConfig, _, err := m.generateContainerConfig(template.container, template.pod, template.attempt, "", template.container.Image, []string{})
```
I think we need to have tests with PodIPs as well across the entire container start logic..

khenidak reviewed (changes requested):
Two general comments:
1. are we sure that `PodIPs` will have values if the gate is turned off? (while we send both PodIP and PodIPs, as we dive deep into call chain we depend only on `PodIP`) -- Maybe we should leave a comment on the need to move everything to PodIP (if we ensure that it will have values even if the gate is off)

2. Please spit commits. One for API/Conversion/Validation and one for kubelet work

aramase commented:
/retest

aojea commented:
SGTM, 
I tested it locally with kind and containerd and the entries are obtained via env variables 

```
MY_POD_IPs=10.244.1.29,fd00:200::1d
MY_POD_IP=10.244.1.29
```

and are present in the /etc/hosts file
```
10.244.1.29     dapi-envars-fieldref
fd00:200::1d    dapi-envars-fieldref
```

aramase commented on pkg/apis/core/pods/helpers.go:
```diff
@@ -88,7 +88,8 @@ func ConvertDownwardAPIFieldLabel(version, label, value string) (string, string,
 		"status.hostIP",
-		"status.podIP":
+		"status.podIP",
+		"status.podIPs":
```
done!

aramase commented:
> are we sure that PodIPs will have values if the gate is turned off? (while we send both PodIP and PodIPs, as we dive deep into call chain we depend only on PodIP) -- Maybe we should leave a comment on the need to move everything to PodIP (if we ensure that it will have values even if the gate is off)

Verified that PodIPs have values even if the gate is turned off. So I've added a note to move everything to use only `PodIPs` instead of `PodIP`.

@khenidak I've updated the PR addressing review comments. PTAL when you get a chance.

aramase commented:
/test pull-kubernetes-integration

aramase commented:
/retest

khenidak commented on pkg/kubelet/container/helpers.go:
```diff
@@ -24,7 +24,7 @@ import (
 	"k8s.io/klog"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
please restore to original value..

aramase replied:
done!

khenidak commented on pkg/kubelet/container/helpers.go:
```diff
@@ -45,7 +45,8 @@ type HandlerRunner interface {
 // able to get necessary informations like the RunContainerOptions, DNS settings, Host IP.
 type RuntimeHelper interface {
-	GenerateRunContainerOptions(pod *v1.Pod, container *v1.Container, podIP string) (contOpts *RunContainerOptions, cleanupAction func(), err error)
+	// TODO (aramase): as podIPs are populated for single stack and dual stack cluster, pass only podIPs.
```
I don't think the `Pod.Status.PodIP` will be ever removed, so that TODO will never be done.

@thockin..

aramase replied:
Removed the TODO.

khenidak commented on pkg/kubelet/kuberuntime/kuberuntime_container_test.go:
```diff
@@ -26,7 +26,7 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
please restore to original value.

aramase replied:
done!

khenidak reviewed (changes requested):
Another review, fewer set of comments. Getting close.

aramase commented:
/test pull-kubernetes-e2e-gce-100-performance

aojea commented:
/test pull-kubernetes-conformance-kind-ipv6

aramase commented:
/assign @thockin @ Random-Liu

aramase commented:
/assign @thockin

andrewsykim commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -128,15 +128,15 @@ func (kl *Kubelet) makeBlockVolumes(pod *v1.Pod, container *v1.Container, podVol
 	// - when the Pod sandbox is being created, its IP is still unknown. Hence, PodIP will not have been set.
-	mountEtcHostsFile := len(podIP) > 0 && runtime.GOOS != "windows"
-	klog.V(3).Infof("container: %v/%v/%v podIP: %q creating hosts mount: %v", pod.Namespace, pod.Name, container.Name, podIP, mountEtcHostsFile)
+	mountEtcHostsFile := len(podIPs) > 0 && runtime.GOOS != "windows"
```
This seems scary, is there a reason we can't keep this as `len(podIP) > 0`? It ensures existing behavior and has the same result since primary pod IP is guaranteed to be set here anyways.

aramase replied:
The `podIP` and `podIPs` are passed from caller `SyncPod`. Here `podIP` is set explicitly to the first value of `podIPs` - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L744-L747. So `podIPs` is guaranteed to be set right? If `podIPs` is empty, then `podIP` is going to be empty here.

lachie83 commented:
/milestone v1.17

thockin reviewed (commented):
Thanks!  Fixes in followups is OK.

/lgtm
/approve

aramase commented on pkg/kubelet/container/helpers.go:
```diff
@@ -45,7 +45,7 @@ type HandlerRunner interface {
 // able to get necessary informations like the RunContainerOptions, DNS settings, Host IP.
 type RuntimeHelper interface {
-	GenerateRunContainerOptions(pod *v1.Pod, container *v1.Container, podIP string) (contOpts *RunContainerOptions, cleanupAction func(), err error)
+	GenerateRunContainerOptions(pod *v1.Pod, container *v1.Container, podIP string, podIPs []string) (contOpts *RunContainerOptions, cleanupAction func(), err error)
```
@thockin correct, podIP is always podsIPs[0]. Will do a followup PR. Thank you for the review.

aramase commented:
/test pull-kubernetes-kubemark-e2e-gce-big
